[RDY] test: win: use cmd.exe by default in job_spec.lua

janlazo opened the pull request:
Ref https://github.com/neovim/neovim/pull/7412#issuecomment-366822555, https://github.com/neovim/neovim/pull/8034
Need #8107 for `jobstop` to return exit code 0 if the program successfully exits and returns exit code 0 if run from a shell (ex. cmd.exe, bash).

Testing to see if switching to cmd.exe improves testing with job API and cat.exe on Windows.

I'm giving up on NULs and newlines for jobsend (list argument). cat.exe still has the `invalid channel id` error, `write_file` truncates the string with `\0` and find.exe truncates the input string after `\0`.

~~`start` (cmd.exe builtin) reaps its child processes with `/b` (I can't find them with tasklist) so I'm using `/min` instead for a minimized window. Redirecting to `nul` shouldn't be necessary. If that's intended with the `jobstop()` test, so be it.~~

Not sure what's the expected behaviour for start (cmd.exe builtin) because nvim is not reaping the child processes on exit after `:qa!`, even on nested nvim termnal. Seems to have different semantics from fork (or what bash does with `&`). Perhaps powershell is better suited here ~~but I don't know how (yes, there's `Start-Process` but I can't pass arguments to the external program).~~ ~~I can use `Start-Job` so if the session terminates, the jobs die too. Is that close enough?~~ Leave it for later. I disabled the pty teardown test.

more.com (like less) is the default pager in Windows so it can be used to display files. That leaves the NUL tests for cat if more can't do it.

fnamemodify doesn't adjust slashes and it's the same on Vim too. Have to use expand() again.

janlazo committed:
test: win: use powershell for Start-Sleep only

cmd.exe (shell) is faster and more reliable than powershell (.NET frontend).
It's best for short and basic tests that don't require non-trivial scripting.
cmd.exe doesn't support sleep so use powershell's Start-Sleep as substitute.

janlazo committed:
test: win: use find.exe /v "" as alternative cat -

justinmk commented:
Somehow the exit code is wrong for these two tests on the 32bit build:

```
[  FAILED  ] C:/projects/neovim/test/functional\core\job_spec.lua @ 140: jobs allows interactive commands
C:/projects/neovim/test/functional\core\job_spec.lua:151: Expected objects to be the same.
Passed in:
(table) {
  [1] = 'notification'
  [2] = 'exit'
 *[3] = {
    [1] = 0
   *[2] = 1 } }
Expected:
(table) {
  [1] = 'notification'
  [2] = 'exit'
 *[3] = {
    [1] = 0
   *[2] = 0 } }
stack traceback:
	C:/projects/neovim/test/functional\core\job_spec.lua:151: in function <C:/projects/neovim/test/functional\core\job_spec.lua:140>
[  FAILED  ] C:/projects/neovim/test/functional\core\job_spec.lua @ 195: jobs preserves NULs
C:/projects/neovim/test/functional\core\job_spec.lua:202: Expected objects to be the same.
Passed in:
(table) {
  [1] = 'notification'
  [2] = 'exit'
 *[3] = {
    [1] = 0
   *[2] = 1 } }
Expected:
(table) {
  [1] = 'notification'
  [2] = 'exit'
 *[3] = {
    [1] = 0
   *[2] = 0 } }
stack traceback:


```

should we leave those "pending" for now?

janlazo commented:
Old tests aren't running. 

```
[00:14:33] C:\projects\neovim\build>cmake --build "C:\projects\neovim\ci\\..\src\nvim\testdir" -- VERBOSE=1 
[00:14:33] Error: could not load cache
[00:14:33] 
[00:14:33] C:\projects\neovim\build>endlocal
```

janlazo commented:
Exit codes are still inconsistent even with cmd.exe but this is more consistent on stdout than powershell.
Powershell sometimes doesn't execute a command if the last argument and `\r\n` aren't separated by whitespace so, when it happens to me, I press Enter again for it to run.

I'll create `cat.cmd` alternative using cmd.exe built-ins to get more consistent exit codes. cat.exe is covered by other tests anyway.

justinmk commented:
Weird. I wonder if we should just write our own `cat.exe` in C, instead of this old gnu binary I found ...

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -138,7 +139,7 @@ describe('jobs', function()
 
   it('allows interactive commands', function()
-    nvim('command', "let j = jobstart(['cat', '-'], g:job_opts)")
+    nvim('command', "let j = jobstart(['"..nvim_cat.."', '-'], g:job_opts)")
```
Doesn't `foo.cmd` require a shell? So I would not expect this to work, the `jobstart({list})` form skips the shell. 

That's one reason why including `cat.exe` was important. 

And changing these invocations to use the shell form of `jobstart({string})` changes the nature of the tests, so we don't want to do that.

janlazo replied:
You're right. I'll keep the list form and run the batchfile with cmd.exe directly. It should be fine if the tempfile doesn't have to be escaped.

janlazo commented:
`find.exe /v ""` (see comments in https://superuser.com/a/853718) is better for multiple newlines. It doesn't preserve NULs though.

janlazo committed:
test: win: try stderr callback and slow output

janlazo commented:
@justinmk How does `jobstop` work in Windows? IIRC, Windows doesn't support signals so killed processes can't clean up.

justinmk commented:
@janlazo Do you mean grandchildren? We will need to iterate through the process tree. 

See `terminate_all` in Vim, which was basically copied from https://stackoverflow.com/q/1173342 .  After a long search, I concluded that's the only way to do it, except another approach used by libuv (see `uv__init_global_job_handle`): create a win32 job object (but that has a quirk in Windows 7 and older: https://stackoverflow.com/a/6138203 ).

The above will be implemented in #8107 .

If you meant something else, please give a specific example.

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -175,63 +168,71 @@ describe('jobs', function()
+    nvim('command', 'call jobsend(j, "'..(iswin()
+      and [[a\n\nc\n\n]]
+      or  [[a\n\nc\n\n\n\nb\n\n]]
+    )..'")')
```
Would rather keep the test pending than have a lot of special-case code that is difficult to follow, just to get the test working.  Maybe something can be improved in the future so that the test can pass without requiring lots of weird workarounds.

janlazo committed:
test: win: enable jobpid() tests

Use ping to test job detach
Use find.exe as an alternative to cat.exe
Use nvim_get_proc to check pid

janlazo commented:
That should be the last of it (unless cat.exe fails again or find.exe errors on `jobstop`).
I can port my changes for powershell from https://github.com/neovim/neovim/pull/8034 to test `jobwait` in the meantime.

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -15,14 +15,25 @@ local expect_twostreams = helpers.expect_twostreams
+  return os.execute((iswin()
+    and 'tasklist /nh /fi "PID eq '..pid..'" | findstr '..pid..' > nul'
+    or  'ps -p '..pid..' > /dev/null'))
+end
```
#8107 provides `nvim_get_proc({pid})` which can be used in the tests via `helpers.meths.get_proc()`.

janlazo replied:
I need to return an exit code though.
Need something like `nvim_get_proc(pid) ~= nil and 1 or 0`.

justinmk replied:
Why do you need an exit code? If `nvim_get_proc` returns `NIL` then the process wasn't found.

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -15,14 +15,25 @@ local expect_twostreams = helpers.expect_twostreams
+  return os.execute((iswin()
+    and 'taskkill /f /t /pid '..pid..' > nul'
+    or  'ps -p '..pid..' > /dev/null'))
+end
```
Would be easy to expose `nvim_kill_proc()` if this turns out to be unreliable.

janlazo committed:
Revert "test/win: use cmd.exe for test"

This reverts commit ae409b5042abdbec67305a063cf921ddffcce0c8.

This PR (#8120) defaults to cmd.exe for job_spec.lua

janlazo committed:
fixup: ping.exe works with cmd.exe pipe

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -179,28 +189,47 @@ describe('jobs', function()
+    nvim('command', "let j = jobstart(has('win32') ? ['find', '/v', ''] : ['cat', '-'], g:job_opts)")
     nvim('command', 'call jobsend(j, "abc\\nxyz")')
-    eq({'notification', 'stdout', {0, {'abc', 'xyz'}}}, next_msg())
+    eq({'notification', 'stdout', {0, iswin() and {'abc', ''} or {'abc', 'xyz'}}}, next_msg())
```
Probably should used expect_msg_seq

janlazo replied:
For this line only? This is expected behaviour for `find.exe` the string has to end with newline (or some character for find.exe receive the input). As for the stdout after jobstop, I don't know why.

I can experiment with `expect_msg_seq` for both stdout.

janlazo replied:
On 2nd thought, I'll revert it. Leave it for `cat` when the situation improves. There's enough tests for find.exe anyway.

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -179,28 +189,47 @@ describe('jobs', function()
+      -- find.exe can't handle 3+ consecutive newlines consistently
+      nvim('command', "let j = jobstart(['find', '/v', ''], g:job_opts)")
+      nvim('command', [[call jobsend(j, "a\n\nc\n\n")]])
+      expect_msg_seq(
```
If it's not consistent then just mark it pending. The entire purpose of this test is to check that newlines are preserved (consistently).

janlazo replied:
I'll revert it back to pending. Revist when testing with cat.exe gets better.

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -215,27 +244,40 @@ describe('jobs', function()
+    end
     nvim('command', "call jobstop(j)")
+    if iswin() then
+      eq({'notification', 'stdout', {0, {'final nl', ''}}}, next_msg())
```
This output has the newline, doesn't it? That defeats the purpose of the test. Should just be pending.

janlazo replied:
It's leftover from stdout when find.exe exits for some reason. Don't recall this happening on a regular cmd.exe window. Strange that this doesn't happen on `jobclose(g:job_id, "stdin")`.

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -260,41 +301,39 @@ describe('jobs', function()
     -- clean up after ourselves
-    os.execute('kill -9 '..pid..' > /dev/null')
+    os_kill(pid)
+    neq(0, os_ps(pid))
```
It should be NOT found since we killed it.

janlazo replied:
I'm  checking if the `os_kill` did work. Is there a better way to do it or do you want to check the exit code of `os_kill`?

justinmk replied:
@janlazo Ah, I didn't realize `os_ps` was returned the exit code, I thought it returned a `pid`. Anyways `nvim_get_proc` should be used instead.

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -511,7 +565,12 @@ describe('jobs', function()
+      \  jobstart('Start-Sleep -Milliseconds 70; exit 4'),
+      \  jobstart('Start-Sleep -Milliseconds 50; exit 5'),
+      \  jobstart('Start-Sleep -Milliseconds 30; exit 6'),
+      \  jobstart('Start-Sleep -Milliseconds 10; exit 7')
```
This is going to be unreliable because powershell startup takes 600ms at least.

janlazo replied:
Bump the sleep timeout to 700, 500, 300, 100 or use cmd.exe and ping instead?
The timeout test for jobwait lasts up to 2-3s so it's long wait either way.

justinmk replied:
I think cmd.exe would be a better choice here.

janlazo replied:
Likely because of Powershell's lazy-loading and Appveyor-specific modules. Can't prevent them with `-NoProfile`.

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -775,22 +840,36 @@ describe("pty process teardown", function()
   end)
 
   it("does not prevent/delay exit. #4798 #4900", function()
```
Could use nvim_get_proc_children instead of a screen test.

janlazo replied:
For the PIDs? It won't work with the child process launched with start (cmd.exe). Have to rethink this test for Windows.

justinmk replied:
Not sure what you mean, but if `!start` is detached from nvim somehow, such that it wouldn't prevent nvim from exiting (if there were a nvim bug in its jobs teardown routine), then that defeats the purpose of the test.

janlazo replied:
nvim will exit completely and `!start` terminates after running the program (ie. ping) but the program that `start` does run (ie ping) will not necessarily terminate if nvim terminates.

justinmk replied:
> but the program that start does run (ie ping) will not necessarily terminate if nvim terminates.

That defeats the purpose of the test :)
